### ISSUE TITLE
Binding output to a specific address still (or again) doesn't work

### ISSUE THREAD

**Issue by grinapo** [syncthing/syncthing#8380]
It seems that the fix for #3682 (https://github.com/syncthing/syncthing/commit/bb23e3940eb9f84134dba3b1741be72c0a8b14bd) was reverted for some reason here: https://github.com/syncthing/syncthing/commit/a27032f09e572dc92fbc69690e519a98164d0634

And of course it has broken it again. Would it be possible to get it back please? :pleading_face: 

Right now it's again impossible to join the relayserver since srv binds to the wrong outbound address. :disappointed:

**Comment by grinapo** [comment]
It seems that this does not use the localAddress, but I don't know go deep enough to see why:

https://github.com/syncthing/syncthing/blob/a27032f09e572dc92fbc69690e519a98164d0634/cmd/strelaysrv/main.go#L143-L146

Also a bit independent but related thing: it would be nice to 
1. log the address used, and
2. log the reject response from the relay server which ought to contain the expected versus the seen IP address.

**Comment by grinapo** [comment]
> Why aren't you just using -ext-address flag?

Because I am only using **one single address** for listen, for outbound (I wish) and it is _the same as the relayserver shall see_. 
In fact I am using the option but it does not change anything at all, I get the same response:

```
strelaysrv -global-rate 10000000 -ext-address 193.227.196.10:22067 -keys /x/syncthing/relay -listen 193.227.196.10:22067 -status-srv 193.227.196.10:22070 -pools https://relays.syncthing.net/endpoint -keys /x/syncthing/relay/../cert -provided-by Lifeforce WHQ

...
Failed to join https://relays.syncthing.net/endpoint: IP address not matching external address
```
And I see that it is indeed using a different address to connect to the relayserver (and goes through a completely unrelated nat gateway, since that is the default outbound for misc services on the machine).

**Comment by calmh** [comment]
I don't think binding to a certain network interface has any effect on the routing of outgoing packets. Linux will follow the main routing table. You probably need to add routing hacks to force traffic out the way you want; this is not something we can control from the application.

**Comment by AudriusButkevicius** [comment]
That's actually very true, I think there is nothing we can do here, hence closing.

**Comment by grinapo** [comment]
@calmh [you are mistaken](https://stackoverflow.com/questions/33768557/how-to-bind-an-http-client-in-go-to-an-ip-address). I have tested that code and it binds the output address to the specified one.

However in my tests (and I repeat that I am not familar with go) **both the linked and your version worked as planned** (and does what you believed to be impossible: binds to an address), at least up to the _transport_ definition:
```golang
  tr := &http.Transpoort{ DialContext: net.Dialer{ LocalAddr: &laddr }.DialContext() }
  // simple test connect (error handling removed for brevity)
  webclient := &http.Client{Transport: tr}
  req, err := http.NewRequest("GET", "http://grin.hu:41808/ip", nil)
  res, err := webclient.Do(req)
```

So it is definitely `LocalAddress` which should do and indeed does the trick. But it seems that somewhere something has been changed and this `tr` doesn't do in the server what it does in my small example based on the linked one. Unfortunately it is rather hard for me to debug it without actual debug output and the lack of deeper golang experience.

It would be extremely nice if this could be fixed. If you have no clue I could try to insert a lot of debug output into the code but it's very time consuming to find the output with trial and error...

**Comment by calmh** [comment]
No, you misunderstand what I'm saying. Certainly we bind to a local address, but this doesn't affect the path the packets take out on the network. The routing is destination based, not source based. If the default route points to a gateway on interface A it doesn't matter if we bind to interface B, the traffic will go out via A regardless (with the source address from B).

**Comment by AudriusButkevicius** [comment]
We don't override the transport in the relaysrv, just set the dialing function. I guess you can try a local build overriding the transport to see if that makes any difference, as thats the only thing that is different.

**Comment by grinapo** [comment]
@calmh oh sorry, I didn't realise you are stating the obvious. ;-)

I didn't detail the topology since it's really irrelevant, and I am sure you're not in dire need to know it either, but in summary: these are containers with a main private address going through a default gw which goes out through a NAT gw, and some containers have a public (ipv4) address which is source-routed to the public gateway, avoiding the NAT but going through a different firewall. (They all have proper, no NAT ipv6 addresses but it seems relay network is ipv4 only. I did not even try to use ipv6-only relay setup, and due to the lack of relevant debug output [or at least my ability to interpret it as relevant] I'm not even sure I could tell whether it worked or not.)

A server running on this container have to bind to a specific publc address, both for listening and for outbound traffic; inbound bind is to prevent traffic directed to non-public or different public address being handled, and outbound is to follow the public [source] routing instead of the NAT one.


@AudriusButkevicius I am not sure what you suggest. I have tried this:
```golang
    localAddr, err := net.ResolveIPAddr("ip", "193.227.196.10")
    localTCPAddr := net.TCPAddr{
        IP: localAddr.IP,
    }

    // synthing ver
    laddr, err := net.ResolveTCPAddr("tcp", "193.227.196.10:12345")
    d := net.Dialer{
        LocalAddr: &localTCPAddr,
    }

    // working version
    tr := &http.Transport{
        Dial:                d.Dial,
    }
    connect(tr)
    
    // syncthing version 
    tr2 := &http.Transport{
        DialContext:          d.DialContext,
    }
    connect(tr2)
    
    laddr.Port = 0;
    d3 := &net.Dialer{ LocalAddr: laddr }
    tr3 := &http.Transport{ DialContext: d3.DialContext }
    connect(tr3)
    
    // empty
    d4 := net.Dialer{}
    tr4 := &http.Transport{
        DialContext:          d4.DialContext,
    }
    connect(tr4)
```
And **all** the first three combos select the proper outbound address, while the last properly using the wrong one. 
I do not see why relaysrv doesn't, but I haven't traced where `tr` gets actually used, and I have realised that strace isn't very helping for golang. If you have no idea where to look I will try to debug it, but, as I mentioned, it's a bit harder for me. But I'm getting more and more curious...
